common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -683,46 +649,14 @@ private static boolean isValidNumericChar(char c) {
      * @return true, if the string represents an IPV4 address in dotted
      *         notation, false otherwise
      */
-    public static boolean isValidIpV4Address(String value) {
-
-        int periods = 0;
-        int i;
-        int length = value.length();
-
-        if (length > 15) {
-            return false;
-        }
-        char c;
-        StringBuilder word = new StringBuilder();
-        for (i = 0; i < length; i ++) {
-            c = value.charAt(i);
-            if (c == '.') {
-                periods ++;
-                if (periods > 3) {
-                    return false;
-                }
-                if (word.length() == 0) {
-                    return false;
-                }
-                if (Integer.parseInt(word.toString()) > 255) {
-                    return false;
-                }
-                word.delete(0, word.length());
-            } else if (!Character.isDigit(c)) {
-                return false;
-            } else {
-                if (word.length() > 2) {
-                    return false;
-                }
-                word.append(c);
-            }
-        }
-
-        if (word.length() == 0 || Integer.parseInt(word.toString()) > 255) {
-            return false;
-        }
-
-        return periods == 3;
+    @SuppressWarnings("DuplicateBooleanBranch")
+    public static boolean isValidIpV4Address(String ip) {
+        int len = ip.length(), i, j;
+        return len <= 15 && len >= 7 &&
+               (i = ip.indexOf('.', 0)) > 0 && isValidIpV4Word(ip, 0, i) &&
+               (i = ip.indexOf('.', j = i + 2)) > 0 && isValidIpV4Word(ip, j - 1, i) &&
+               (i = ip.indexOf('.', j = i + 2)) > 0 && isValidIpV4Word(ip, j - 1, i) &&
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Identical sub-expressions on both sides of operator "&&" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1764)
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Identical sub-expressions on both sides of operator "&&" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1764)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract the assignment out of this expression. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)